File: csharp/tests/BindingTests.cs
/*
 * $Id$
 *
 * C# unit tests for Lasso library
 *
 * Copyright (C) 2004, 2005 Entr'ouvert
 * http://lasso.entrouvert.org
 *
 * Authors: See AUTHORS file.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* 
 * To run it, install Lasso then compile with:
 * export PKG_CONFIG_PATH=../
 * ln -s ../lasso.dll
 * ln -s ../lasso.dll.config
 * mcs -g -nologo -pkg:lasso-sharp -out:BindingTests.exe BindingTests.cs
 */

using System;

public class BindingTests {
	static void assertEquals(int i1, int i2) {
		if (i1 != i2)
			Console.WriteLine("Assertion failed: %d != %d", i1, i2);
	}

	static void assertEquals(String s1, String s2) {
		if (s1 != s2)
			Console.WriteLine("Assertion failed: %s != %s", s1, s2);
	}

	static void assertNull(Object o) {
		if (o != null)
			Console.WriteLine("Assertion failed: %s is not null", o);
	}

	static void assertNull(String s) {
		if (s != null)
			Console.WriteLine("Assertion failed: %s is not null", s);
	}

	static void Main() {
		lasso.lasso.init();
		test01();
		test02();
		test03();
		test04();
		test05();
		test06();
		lasso.lasso.shutdown();
	}

	static void test01() {
		Console.Write("Create and delete nodes.");

		lasso.LibAuthnRequest authnRequest = new lasso.LibAuthnRequest();
		authnRequest = null;

		Console.WriteLine(".. OK");
	}

	static void test02() {
		Console.Write("Get & set simple attributes of nodes.");

		lasso.LibAuthnRequest authnRequest = new lasso.LibAuthnRequest();

		// Test a string attribute.
		assertNull(authnRequest.consent);
		authnRequest.consent = lasso.lasso.libConsentObtained;
		assertEquals(authnRequest.consent, lasso.lasso.libConsentObtained);
		authnRequest.consent = null;
		assertNull(authnRequest.consent);

		// Test a renamed string attribute.
		assertNull(authnRequest.relayState);
		authnRequest.relayState = "Hello World!";
		assertEquals(authnRequest.relayState, "Hello World!");
		authnRequest.relayState = null;
		assertNull(authnRequest.relayState);

		// Test an integer attribute.
		assertEquals(authnRequest.majorVersion, 0);
		authnRequest.majorVersion = 314;
		assertEquals(authnRequest.majorVersion, 314);

		authnRequest = null;

		Console.WriteLine(".. OK");
	}

	static void test03() {
		Console.Write("Get & set attributes of nodes of type string list.");

		lasso.LibAuthnRequest authnRequest = new lasso.LibAuthnRequest();

		assertNull(authnRequest.respondWith);

		lasso.StringList respondWith = new lasso.StringList();
		assertEquals(respondWith.length(), 0);
		respondWith.append("first string");
		assertEquals(respondWith.length(), 1);
		assertEquals(respondWith.getItem(0), "first string");
		assertEquals(respondWith.getItem(0), "first string");
		respondWith.append("second string");
		assertEquals(respondWith.length(), 2);
		assertEquals(respondWith.getItem(0), "first string");
		assertEquals(respondWith.getItem(1), "second string");
		respondWith.append("third string");
		assertEquals(respondWith.length(), 3);
		assertEquals(respondWith.getItem(0), "first string");
		assertEquals(respondWith.getItem(1), "second string");
		assertEquals(respondWith.getItem(2), "third string");
		authnRequest.respondWith = respondWith;
		assertEquals(authnRequest.respondWith.getItem(0), "first string");
		assertEquals(authnRequest.respondWith.getItem(1), "second string");
		assertEquals(authnRequest.respondWith.getItem(2), "third string");
		assertEquals(respondWith.getItem(0), "first string");
		assertEquals(respondWith.getItem(1), "second string");
		assertEquals(respondWith.getItem(2), "third string");
		respondWith = null;
		assertEquals(authnRequest.respondWith.getItem(0), "first string");
		assertEquals(authnRequest.respondWith.getItem(1), "second string");
		assertEquals(authnRequest.respondWith.getItem(2), "third string");
		respondWith = authnRequest.respondWith;
		assertEquals(respondWith.getItem(0), "first string");
		assertEquals(respondWith.getItem(1), "second string");
		assertEquals(respondWith.getItem(2), "third string");
		respondWith = null;
		assertEquals(authnRequest.respondWith.getItem(0), "first string");
		assertEquals(authnRequest.respondWith.getItem(1), "second string");
		assertEquals(authnRequest.respondWith.getItem(2), "third string");
		authnRequest.respondWith = null;
		assertNull(authnRequest.respondWith);
	
		authnRequest = null;

		Console.WriteLine(".. OK");
	}

	static void test04() {
		Console.Write("Get & set attributes of nodes of type node list.");

	        lasso.SamlpResponse response = new lasso.SamlpResponse();

	        assertNull(response.assertion);
	        lasso.NodeList assertions = new lasso.NodeList();
	        assertEquals(assertions.length(), 0);
	        lasso.SamlAssertion assertion1 = new lasso.SamlAssertion();
	        assertion1.assertionId = "assertion 1";
	        assertions.append(assertion1);
	        assertEquals(assertions.length(), 1);
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        lasso.SamlAssertion assertion2 = new lasso.SamlAssertion();
	        assertion2.assertionId = "assertion 2";
	        assertions.append(assertion2);
	        assertEquals(assertions.length(), 2);
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(1)).assertionId,
				"assertion 2");
	        lasso.SamlAssertion assertion3 = new lasso.SamlAssertion();
	        assertion3.assertionId = "assertion 3";
	        assertions.append(assertion3);
	        assertEquals(assertions.length(), 3);
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(2)).assertionId,
				"assertion 3");
	        response.assertion = assertions;
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(2)).assertionId,
				"assertion 3");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(2)).assertionId,
				"assertion 3");
	        assertions = null;;
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(2)).assertionId,
				"assertion 3");
	        assertions = response.assertion;
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) assertions.getItem(2)).assertionId,
				"assertion 3");
	        assertions = null;
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(0)).assertionId,
				"assertion 1");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(1)).assertionId,
				"assertion 2");
	        assertEquals(((lasso.SamlAssertion) response.assertion.getItem(2)).assertionId,
				"assertion 3");
	        response.assertion = null;
	        assertNull(response.assertion);

		response = null;

		Console.WriteLine(".. OK");
	}

	static void test05() {
		Console.Write("Get & set attributes of nodes of type XML list.");

		lasso.LibAuthnRequest authnRequest = new lasso.LibAuthnRequest();

		assertNull(authnRequest.extension);

	        String actionString1 = "<lib:Extension xmlns:lib=\"urn:liberty:iff:2003-08\">\n"
			+ "  <action>do 1</action>\n"
			+ "</lib:Extension>";
	        String actionString2 = "<lib:Extension xmlns:lib=\"urn:liberty:iff:2003-08\">\n"
			+ "  <action>do 2</action>\n"
			+ "</lib:Extension>";
	        String actionString3 = "<lib:Extension xmlns:lib=\"urn:liberty:iff:2003-08\">\n"
			+ "  <action>do 3</action>\n"
			+ "</lib:Extension>";
		lasso.StringList extension = new lasso.StringList();
		assertEquals(extension.length(), 0);
		extension.append(actionString1);
		assertEquals(extension.length(), 1);
		assertEquals(extension.getItem(0), actionString1);
		assertEquals(extension.getItem(0), actionString1);
		extension.append(actionString2);
		assertEquals(extension.length(), 2);
		assertEquals(extension.getItem(0), actionString1);
		assertEquals(extension.getItem(1), actionString2);
		extension.append(actionString3);
		assertEquals(extension.length(), 3);
		assertEquals(extension.getItem(0), actionString1);
		assertEquals(extension.getItem(1), actionString2);
		assertEquals(extension.getItem(2), actionString3);
		authnRequest.extension = extension;
		assertEquals(authnRequest.extension.getItem(0), actionString1);
		assertEquals(authnRequest.extension.getItem(1), actionString2);
		assertEquals(authnRequest.extension.getItem(2), actionString3);
		assertEquals(extension.getItem(0), actionString1);
		assertEquals(extension.getItem(1), actionString2);
		assertEquals(extension.getItem(2), actionString3);
		extension = null;
		assertEquals(authnRequest.extension.getItem(0), actionString1);
		assertEquals(authnRequest.extension.getItem(1), actionString2);
		assertEquals(authnRequest.extension.getItem(2), actionString3);
		extension = authnRequest.extension;
		assertEquals(extension.getItem(0), actionString1);
		assertEquals(extension.getItem(1), actionString2);
		assertEquals(extension.getItem(2), actionString3);
		extension = null;
		assertEquals(authnRequest.extension.getItem(0), actionString1);
		assertEquals(authnRequest.extension.getItem(1), actionString2);
		assertEquals(authnRequest.extension.getItem(2), actionString3);
		authnRequest.extension = null;
		assertNull(authnRequest.extension);

		authnRequest = null;

		Console.WriteLine(".. OK");
	}

	static void test06() {
		Console.Write("Get & set attributes of nodes of type node.");

	        lasso.Login login = new lasso.Login(new lasso.Server(null, null, null, null));

	        assertNull(login.request);
	        login.request = (lasso.SamlpRequestAbstract) new lasso.LibAuthnRequest();
	        ((lasso.LibAuthnRequest) login.request).consent = lasso.lasso.libConsentObtained;
	        assertEquals(((lasso.LibAuthnRequest) login.request).consent,
			 lasso.lasso.libConsentObtained);
	        login.request = null;
	        assertNull(login.request);
	
	        login = null;

		Console.WriteLine(".. OK");
	}
}

File: perl/tests/binding_tests.pl
#!/usr/bin/env perl

use lasso;

sub test01 {
    print "Create and delete nodes.";

    my $authnRequest = new lasso::LibAuthnRequest;
    undef $authnRequest;

    print "..OK\n";
}

sub test02 {
    print "Get & set simple attributes of nodes.";

    my $authnRequest = new lasso::LibAuthnRequest;

    # Test a string attribute.
    $authnRequest->{consent} eq undef or die "Assertion failed\n";
    $authnRequest->{consent} = $lasso::libConsentObtained;
    $authnRequest->{consent} eq $lasso::libConsentObtained or die "Assertion failed\n";
    $authnRequest->{consent} = undef;
    $authnRequest->{consent} eq undef or die "Assertion failed\n";

    # Test a renamed string attribute.
    $authnRequest->{relayState} eq undef or die "Assertion failed\n";
    $authnRequest->{relayState} = 'Hello World!';
    $authnRequest->{relayState} eq 'Hello World!' or die "Assertion failed\n";
    $authnRequest->{relayState} = undef;
    $authnRequest->{relayState} eq undef or die "Assertion failed\n";

    # Test an integer attribute.
    $authnRequest->{majorVersion} eq 0 or die "Assertion failed\n";
    $authnRequest->{majorVersion} = 314;
    $authnRequest->{majorVersion} eq 314 or die "Assertion failed\n";

    undef $authnRequest;

    print "..OK\n";
}

sub test03 {
    print "Get & set attributes of nodes of type string list.";

    my $authnRequest = new lasso::LibAuthnRequest;

    $authnRequest->{respondWith} eq undef or die "Assertion failed\n";

    my $respondWith = new lasso::StringList;
    $respondWith->length eq 0 or die "Assertion failed\n";
    $respondWith->append('first string');
    $respondWith->length eq 1 or die "Assertion failed\n";
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->append('second string');
    $respondWith->length eq 2 or die "Assertion failed\n";
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->getitem(1) eq 'second string' or die "Assertion failed\n";
    $respondWith->append('third string');
    $respondWith->length eq 3 or die "Assertion failed\n";
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->getitem(1) eq 'second string' or die "Assertion failed\n";
    $respondWith->getitem(2) eq 'third string' or die "Assertion failed\n";
    $authnRequest->{respondWith} = $respondWith;
    $authnRequest->{respondWith}->getitem(0) eq 'first string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(1) eq 'second string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(2) eq 'third string' or die "Assertion failed\n";
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->getitem(1) eq 'second string' or die "Assertion failed\n";
    $respondWith->getitem(2) eq 'third string' or die "Assertion failed\n";
    undef $respondWith;
    $authnRequest->{respondWith}->getitem(0) eq 'first string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(1) eq 'second string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(2) eq 'third string' or die "Assertion failed\n";
    $respondWith = $authnRequest->{respondWith};
    $respondWith->getitem(0) eq 'first string' or die "Assertion failed\n";
    $respondWith->getitem(1) eq 'second string' or die "Assertion failed\n";
    $respondWith->getitem(2) eq 'third string' or die "Assertion failed\n";
    undef $respondWith;
    $authnRequest->{respondWith}->getitem(0) eq 'first string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(1) eq 'second string' or die "Assertion failed\n";
    $authnRequest->{respondWith}->getitem(2) eq 'third string' or die "Assertion failed\n";
    $authnRequest->{respondWith} = undef;
    $authnRequest->{respondWith} eq undef or die "Assertion failed\n";

    undef $authnRequest;

    print "..OK\n";
}

sub test04 {
    print "Get & set attributes of nodes of type node list.";

    my $response = new lasso::SamlpResponse;

    $response->{assertion} eq undef or die "Assertion failed\n";

    my $assertions = new lasso::NodeList;
    $assertions->length eq 0 or die "Assertion failed\n";
    my $assertion1 = new lasso::SamlAssertion;
    $assertion1->{assertionId} = 'assertion 1';
    $assertions->append($assertion1);
    $assertions->length eq 1 or die "Assertion failed\n";
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    my $assertion2 = new lasso::SamlAssertion;
    $assertion2->{assertionId} = 'assertion 2';
    $assertions->append($assertion2);
    $assertions->length eq 2 or die "Assertion failed\n";
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $assertions->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    my $assertion3 = new lasso::SamlAssertion;
    $assertion3->{assertionId} = 'assertion 3';
    $assertions->append($assertion3);
    $assertions->length eq 3 or die "Assertion failed\n";
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $assertions->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $assertions->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    $response->{assertion} = $assertions;
    $response->{assertion}->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $response->{assertion}->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $response->{assertion}->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $assertions->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $assertions->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    undef $assertions;
    $response->{assertion}->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $response->{assertion}->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $response->{assertion}->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    $assertions = $response->{assertion};
    $assertions->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $assertions->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $assertions->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    undef $assertions;
    $response->{assertion}->getitem(0)->{assertionId} eq 'assertion 1' or die "Assertion failed\n";
    $response->{assertion}->getitem(1)->{assertionId} eq 'assertion 2' or die "Assertion failed\n";
    $response->{assertion}->getitem(2)->{assertionId} eq 'assertion 3' or die "Assertion failed\n";
    $response->{assertion} = undef;
    $response->{assertion} eq undef or die "Assertion failed\n";

    undef $response;

    print "..OK\n";
}

sub test05 {
    print "Get & set attributes of nodes of type XML list.";

    my $authnRequest = new lasso::LibAuthnRequest;

    $authnRequest->{extension} eq undef or die "Assertion failed\n";

    my $actionString1 = '<lib:Extension xmlns:lib="urn:liberty:iff:2003-08">
  <action>do 1</action>
</lib:Extension>';
    my $actionString2 = '<lib:Extension xmlns:lib="urn:liberty:iff:2003-08">
  <action>do 2</action>
</lib:Extension>';
    my $actionString3 = '<lib:Extension xmlns:lib="urn:liberty:iff:2003-08">
  <action>do 3</action>
</lib:Extension>';
    my $extension = new lasso::StringList;
    $extension->length eq 0 or die "Assertion failed\n";
    $extension->append($actionString1);
    $extension->length eq 1 or die "Assertion failed\n";
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->append($actionString2);
    $extension->length eq 2 or die "Assertion failed\n";
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $extension->append($actionString3);
    $extension->length eq 3 or die "Assertion failed\n";
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $extension->getitem(2) eq $actionString3 or die "Assertion failed\n";
    $authnRequest->{extension} = $extension;
    $authnRequest->{extension}->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(2) eq $actionString3 or die "Assertion failed\n";
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $extension->getitem(2) eq $actionString3 or die "Assertion failed\n";
    undef $extension;
    $authnRequest->{extension}->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(2) eq $actionString3 or die "Assertion failed\n";
    $extension = $authnRequest->{extension};
    $extension->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $extension->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $extension->getitem(2) eq $actionString3 or die "Assertion failed\n";
    undef $extension;
    $authnRequest->{extension}->getitem(0) eq $actionString1 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(1) eq $actionString2 or die "Assertion failed\n";
    $authnRequest->{extension}->getitem(2) eq $actionString3 or die "Assertion failed\n";
    $authnRequest->{extension} = undef;
    $authnRequest->{extension} eq undef or die "Assertion failed\n";

    undef $authnRequest;

    print "..OK\n";
}

sub test06 {
    print "Get & set attributes of nodes of type node.";

    my $login = new lasso::Login(new lasso::Server);

    $login->{request} eq undef or die "Assertion failed\n";
    $login->{request} = new lasso::LibAuthnRequest;
    $login->{request}->{consent} = lasso::libConsentObtained;
    $login->{request}->{consent} eq lasso::libConsentObtained or die "Assertion failed\n";
    undef $login->{request};
    $login->{request} eq undef or die "Assertion failed\n";

    undef $login;

    print "..OK\n";
}

lasso::init;
test01;
test02;
test03;
test04;
test05;
test06;
lasso::shutdown();
